File: calculator/app/src/main/java/com/example/calculator/MainActivity.java
package com.example.calculator;


import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    TextView input;
    String sign, value1, value2;
    Double num1, num2, result;
    boolean hasDot;
    boolean hasoper=false;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        input = findViewById(R.id.input);
        if(savedInstanceState!=null) {
            String firts,number1,number2,oper;
            boolean dot,haser;
            input.setText(savedInstanceState.getString("first_Value"));
            value1=savedInstanceState.getString("number1");
            sign=savedInstanceState.getString("operation");
            value2=savedInstanceState.getString("number2");
            hasDot=savedInstanceState.getBoolean("dot");
            hasoper=savedInstanceState.getBoolean("haser");
        }

    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("first_Value",input.getText().toString());
        outState.putString("number1",value1);
        outState.putString("operation",sign);
        outState.putString("number2",value2);
        outState.putBoolean("dot",hasDot);
        outState.putBoolean("haser",hasoper);


    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        savedInstanceState.getString("first_Value");
        savedInstanceState.getString("number1");
        savedInstanceState.getString("operation");
        savedInstanceState.getString("number2");
        savedInstanceState.getBoolean("dot");
        savedInstanceState.getBoolean("haser");

    }

    public void btnClick_0(View view) {
        String s=input.getText().toString();
        if( s.equals("0")){

        }
        else{
            input.setText(input.getText()+"0");
        }
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"0");
        }

    }
    public void btnClick_1(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"1");
        }
        else {
            if (s.equals("0")) {
                input.setText(null);
                input.setText(input.getText() + "1");
            } else {
                input.setText((input.getText() + "1"));
            }
        }
    }
    public void btnClick_2(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"2");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText(input.getText()+"2");}
            else{
                input.setText((input.getText()+"2"));
            }}
    }


    public void btnClick_3(View view){
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText("3");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText("3");
            }
            else{
                input.setText(input.getText()+"3");
            }
        }

    }

    public void btnClick_4(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"4");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText(input.getText()+"4");}
            else{
                input.setText((input.getText()+"4"));
            }}

    }
    public void btnClick_5(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"5");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText(input.getText()+"5");}
            else{
                input.setText((input.getText()+"5"));
            }}
    }
    public void btnClick_6(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"6");
        }
        else {
            if (s.equals("0")) {
                input.setText(null);
                input.setText(input.getText() + "6");
            } else {
                input.setText((input.getText() + "6"));
            }
        }
    }
    public void btnClick_7(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"7");
        }
        else {
            if (s.equals("0")) {
                input.setText(null);
                input.setText(input.getText() + "7");
            } else {
                input.setText((input.getText() + "7"));
            }
        }
    }
    public void btnClick_8(View view) {
        String s=input.getText().toString();
        if(s.equals("math error")){
            input.setText(null);
            input.setText(input.getText()+"8");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText(input.getText()+"8");}
            else{
                input.setText((input.getText()+"8"));
            }}

    }
    public void btnClick_9(View view) {
        String s=input.getText().toString();

        if(s.equals("math error")){
            input.setText(null);
            input.setText("9");
        }
        else{
            if(s.equals("0")){
                input.setText(null);
                input.setText(input.getText()+"9");}
            else{
                input.setText((input.getText()+"9"));
            }
        }
    }


    public void btnClick_plus(View view) {
        if(!hasoper){
            sign = "+";
            value1 = input.getText().toString();
            input.setText(null);
            hasoper=true;
            hasDot = false;}
        else{
            sign="+";
        }
    }

    public void btnClick_minus(View view) {
        if(!hasoper){
            sign = "-";
            value1 = input.getText().toString();
            input.setText(null);
            hasoper=true;
            hasDot = false;}
        else {
            sign="-";
        }
    }

    public void btnClick_multip(View view) {
        if(!hasoper){
            sign = "*";
            value1 = input.getText().toString();
            input.setText(null);
            hasoper=true;
            hasDot = false;}
        else{
            sign="*";
        }
    }

    public void btnClick_devid(View view) {
        String s = input.getText().toString();
        if (s.equals("math error")) {
            input.setText("math error");
        } else {
            if (!hasoper) {
                sign = "/";
                value1 = input.getText().toString();
                hasoper = true;
                input.setText(null);
                hasDot = false;
            } else {
                sign = "/";
            }
        }
    }

    public void btnClick_dota(View view) {
        if (!hasDot) {
            if (input.getText().equals("")) {

                input.setText("0.");
            } else {

                input.setText(input.getText() + ".");
            }

            hasDot = true;
        }
    }

    public void btnClick_koren(View view) {

        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double koren = Double.parseDouble(s);
        input.setText(Math.sqrt(koren) + "");
        hasDot = true;

    }

    public void btnClick_kvadr(View view) {
        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double t= Double.parseDouble(s);
        input.setText(Math.pow(t,2)+"");
        hasDot=true;
    }

    public void btnClick_clear(View view) {
        input.setText(null);
        value1 = null;
        value2 = null;
        sign = null;
        hasDot = false;
    }


    public void btnClick_delete (View view){

        String s=input.getText().toString();
        if(s.equals(""))
        {

        }
        else {
            int t = s.length();
            String s1 = s.substring(0, t - 1);
            input.setText(s1 + "");
        }
    }
    public void btnClick_rovno(View view) {
        if (sign == null) {
            {

            }
        } else if (input.getText().equals("")) {
            {}
        }
        else {
            switch (sign) {
                default:
                    break;
                case "+":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 + num2;
                    input.setText(result + "");
                    hasDot=true;
                    hasoper=false;
                    sign = null;
                    break;
                case "-":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 - num2;
                    input.setText(result + "");
                    hasDot=true;
                    hasoper=false;
                    sign = null;
                    break;
                case "*":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 * num2;
                    input.setText(result + "");
                    hasDot=true;
                    hasoper=false;
                    sign = null;
                    break;
                case "/":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    value2 = input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = num1 / num2;
                    if(result.toString().equals("Infinity") || result.toString().equals("NaN")){
                        input.setText("math error");
                    }
                    else {
                        input.setText(result + "");}
                    hasDot=true;
                    sign = null;
                    hasoper=false;
                    break;
                case "kor":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    value2=input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = Math.pow(Math.E, Math.log(num1)/num2);
                    input.setText(result + "");
                    hasDot=true;
                    sign = null;
                    hasoper=false;
                    break;
                case "^":
                    if(value1=="" || value1.equals("Infinit") || value1.equals("Infini") || value1.equals("Infin") || value1.equals("Infi") || value1.equals("Inf") || value1.equals("In") || value1.equals("I") || value1.equals("math error")){value1="0";}
                    int t;
                    value2=input.getText().toString();
                    num1 = Double.parseDouble(value1);
                    num2 = Double.parseDouble(value2);
                    result = Math.pow(num1,(num2));
                    input.setText(result + "");
                    hasDot=true;
                    sign = null;
                    hasoper=false;
                    break;
            }

        }
    }

    public void btnClick_sin(View view) {
        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double f= Double.parseDouble(s);
        input.setText(Math.cos(f)+"");
    }

    public void btnClick_cos(View view) {
        String s=input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double f=Double.parseDouble(s);
        input.setText(Math.cos(f)+"");
    }

    public void btnClick_tan(View view) {
        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double f= Double.parseDouble(s);
        input.setText(Math.tan(f)+"");
    }

    public void btnClick_persent(View view) {
        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double f= Double.parseDouble(s);
        f=f*0.01;
        input.setText(f+"");
    }

    public void btnClick_ln(View view) {
        String s= input.getText().toString();
        if(s.equals("")){

        }else{
            if(s.equals("math error"))
            {
                s = "0";
            }
            double f= Double.parseDouble(s);
            f=Math.log(f);
            input.setText(f+"");}
    }

    public void btnClick_log(View view) {
        String s= input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        double f= Double.parseDouble(s);
        f=Math.log10(f);
        input.setText(f+"");
    }


    public void btnClick_ndar(View view) {
        String s = input.getText().toString();
        if (s.equals("math error")) {
            input.setText("math error");
        } else {
            if (!hasoper) {
                sign = "^";
                value1 = input.getText().toString();
                hasoper = true;
                input.setText(null);
                hasDot = false;
            } else {
                sign = "^";
            }
        }
    }

    public void btnClick_nkoren(View view) {
        String s=input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        value1=s;
        if(!hasoper){
            sign="kor";
            value1=input.getText().toString();
            hasoper=true;
            input.setText(null);
            hasDot=false;
        }
        else{
            sign="kor";
        }

    }

    public void btnClick_factorial(View view) {

        String s=input.getText().toString();
        if(s.equals("math error"))
        {
            s = "0";
        }
        boolean ok = true;
        long a = Long.parseLong(s);
        if((a == 1) || a == 0)
        {
            a = 1;
            input.setText(a+"");
        }
        else
        {
            if(a > 60)
            {
                input.setText("math error");
            }
            else
            {
                double fact = 1;
                for(int i = 2; i <= a; i++)
                {
                    fact = fact * i;
                }
                input.setText(fact+"");
            }
        }


    }
}

File: calculator_na_kotline/app/src/main/java/com/example/calculator_na_kotline/MainActivity.kt
package com.example.calculator_na_kotline

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var input: TextView
    internal var sign: String? = null
    internal var value1: String? = null
    internal var value2: String? = null
    internal var num1: Double? = null
    internal var num2: Double? = null
    internal var result: Double? = null
    internal var hasDot: Boolean = false
    internal var hasoper: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        input = findViewById(R.id.input)
        if(savedInstanceState != null){
            value1 = savedInstanceState.getString("number1")
            value2 = savedInstanceState.getString("number2")
            sign = savedInstanceState.getString("sign")
            hasDot = savedInstanceState.getBoolean("dot")
            hasoper = savedInstanceState.getBoolean("haser")
            input.text = savedInstanceState.getString("first_value")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("first_Value", input.text.toString())
        outState.putString("number1", value1)
        outState.putString("operation", sign)
        outState.putString("number2", value2)
        outState.putBoolean("dot", hasDot)
        outState.putBoolean("haser", hasoper)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        savedInstanceState.getString("first_Value")
        savedInstanceState.getString("number1")
        savedInstanceState.getString("operation")
        savedInstanceState.getString("number2")
        savedInstanceState.getBoolean("dot")
        savedInstanceState.getBoolean("haser")
    }

    fun btnClick_0(view: View?) {
        val s = input.text.toString()
        if (s == "0") {
        } else {
            input.text = input.text.toString() + "0"
        }
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "0"
        }
    }

    fun btnClick_1(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "1"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "1"
            } else {
                input.text = input.text.toString() + "1"
            }
        }
    }

    fun btnClick_2(view: View?) {

        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "2"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "2"
            } else {
                input.text = input.text.toString() + "2"
            }
        }
    }

    fun btnClick_3(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "3"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "3"
            } else {
                input.text = input.text.toString() + "3"
            }
        }
    }

    fun btnClick_4(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "4"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "4"
            } else {
                input.text = input.text.toString() + "4"
            }
        }
    }

    fun btnClick_5(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "5"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "5"
            } else {
                input.text = input.text.toString() + "5"
            }
        }
    }

    fun btnClick_6(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "6"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "6"
            } else {
                input.text = input.text.toString() + "6"
            }
        }
    }

    fun btnClick_7(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "7"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "7"
            } else {
                input.text = input.text.toString() + "7"
            }
        }
    }

    fun btnClick_8(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "8"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "8"
            } else {
                input.text = input.text.toString() + "8"
            }
        }
    }

    fun btnClick_9(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = null
            input.text = input.text.toString() + "9"
        } else {
            if (s == "0") {
                input.text = null
                input.text = input.text.toString() + "9"
            } else {
                input.text = input.text.toString() + "9"
            }
        }
    }

    fun btnClick_plus(view: View?) {
        if (!hasoper) {
            sign = "+"
            value1 = input.text.toString()
            input.text = null
            hasoper = true
            hasDot = false
        } else {
            sign = "+"
        }
    }

    fun btnClick_minus(view: View?) {
        if (!hasoper) {
            sign = "-"
            value1 = input.text.toString()
            input.text = null
            hasoper = true
            hasDot = false
        } else {
            sign = "-"
        }
    }

    fun btnClick_multip(view: View?) {
        if (!hasoper) {
            sign = "*"
            value1 = input.text.toString()
            input.text = null
            hasoper = true
            hasDot = false
        } else {
            sign = "*"
        }
    }

    fun btnClick_devid(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = "math error"
        } else {
            if (!hasoper) {
                sign = "/"
                value1 = input.text.toString()
                hasoper = true
                input.text = null
                hasDot = false
            } else {
                sign = "/"
            }
        }
    }

    fun btnClick_dota(view: View?) {
        if (!hasDot) {
            if (input.text == "") {
                input.text = "0."
            } else {
                input.text = input.text.toString() + "."
            }
            hasDot = true
        }
    }

    fun btnClick_koren(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val koren = s.toDouble()
        input.text = Math.sqrt(koren).toString() + ""
        hasDot = true
    }

    fun btnClick_kvadr(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val t = s.toDouble()
        input.text = Math.pow(t, 2.0).toString() + ""
        hasDot = true
    }

    fun btnClick_clear(view: View?) {
        input.text = null
        value1 = null
        value2 = null
        sign = null
        hasDot = false
    }


    fun btnClick_delete(view: View?) {
        val s = input.text.toString()
        if (s == "") {
        } else {
            val t = s.length
            val s1 = s.substring(0, t - 1)
            input.text = s1 + ""
        }
    }

    fun btnClick_rovno(view: View?) {
        if (sign == null) {
            run {}
        } else if (input.text == "") {
            run {}
        } else {
            when (sign) {
                "+" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = num1!! + num2!!
                    input.text = result.toString() + ""
                    hasDot = true
                    hasoper = false
                    sign = null
                }
                "-" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = num1!! - num2!!
                    input.text = result.toString() + ""
                    hasDot = true
                    hasoper = false
                    sign = null
                }
                "*" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = num1!! * num2!!
                    input.text = result.toString() + ""
                    hasDot = true
                    hasoper = false
                    sign = null
                }
                "/" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = num1!! / num2!!
                    if (result.toString() == "Infinity" || result.toString() == "NaN") {
                        input.text = "math error"
                    } else {
                        input.text = result.toString() + ""
                    }
                    hasDot = true
                    sign = null
                    hasoper = false
                }
                "kor" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = Math.pow(Math.E, Math.log(num1!!) / num2!!)
                    input.text = result.toString() + ""
                    hasDot = true
                    sign = null
                    hasoper = false
                }
                "^" -> {
                    if (value1 === "" || value1 == "Infinit" || value1 == "Infini" || value1 == "Infin" || value1 == "Infi" || value1 == "Inf" || value1 == "In" || value1 == "I" || value1 == "math error") {
                        value1 = "0"
                    }
                    var t: Int
                    value2 = input.text.toString()
                    num1 = value1!!.toDouble()
                    num2 = value2!!.toDouble()
                    result = Math.pow(num1!!, num2!!)
                    input.text = result.toString() + ""
                    hasDot = true
                    sign = null
                    hasoper = false
                }
                else -> {
                }
            }
        }
    }

    fun btnClick_sin(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val f = s.toDouble()
        input.text = Math.cos(f).toString() + ""
    }

    fun btnClick_cos(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val f = s.toDouble()
        input.text = Math.cos(f).toString() + ""
    }

    fun btnClick_tan(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val f = s.toDouble()
        input.text = Math.tan(f).toString() + ""
    }

    fun btnClick_persent(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        var f = s.toDouble()
        f = f * 0.01
        input.text = f.toString() + ""
    }

    fun btnClick_ln(view: View?) {
        var s = input.text.toString()
        if (s == "") {
        } else {
            if (s == "math error") {
                s = "0"
            }
            var f = s.toDouble()
            f = Math.log(f)
            input.text = f.toString() + ""
        }
    }

    fun btnClick_log(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        var f = s.toDouble()
        f = Math.log10(f)
        input.text = f.toString() + ""
    }


    fun btnClick_ndar(view: View?) {
        val s = input.text.toString()
        if (s == "math error") {
            input.text = "math error"
        } else {
            if (!hasoper) {
                sign = "^"
                value1 = input.text.toString()
                hasoper = true
                input.text = null
                hasDot = false
            } else {
                sign = "^"
            }
        }
    }

    fun btnClick_nkoren(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        value1 = s
        if (!hasoper) {
            sign = "kor"
            value1 = input.text.toString()
            hasoper = true
            input.text = null
            hasDot = false
        } else {
            sign = "kor"
        }
    }

    fun btnClick_factorial(view: View?) {
        var s = input.text.toString()
        if (s == "math error") {
            s = "0"
        }
        val ok = true
        var a = s.toLong()
        if (a == 1L || a == 0L) {
            a = 1
            input.text = a.toString() + ""
        } else {
            if (a > 60) {
                input.text = "math error"
            } else {
                var fact = 1.0
                for (i in 2..a) {
                    fact = fact * i
                }
                input.text = fact.toString() + ""
            }
        }
    }

}
